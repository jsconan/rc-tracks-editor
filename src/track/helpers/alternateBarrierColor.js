/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { alternate } from '../../core/helpers';

/**
 * A list of colors for the barriers.
 * @type {string[]}
 */
const barrierColors = ['even', 'odd'];

/**
 * Returns the color for a barrier element at the given position.
 * @param {number} i - The position of the barrier element.
 * @returns {string} - The color for the barrier element.
 */
export default i => alternate(i, barrierColors);
