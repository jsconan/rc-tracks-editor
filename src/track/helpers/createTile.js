/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {
    validateType,
    CURVED_TILE_ENLARGED_TYPE,
    CURVED_TILE_TYPE,
    STRAIGHT_TILE_TYPE,
    TILE_DIRECTION_RIGHT
} from '../../tile/helpers';
import { CurvedTileEnlargedModel, CurvedTileModel, StraightTileModel } from '../../tile/models';

/**
 * @type {object} - Maps the types of tile to their respective model.
 * @private
 */
const modelsMap = {
    [STRAIGHT_TILE_TYPE]: StraightTileModel,
    [CURVED_TILE_TYPE]: CurvedTileModel,
    [CURVED_TILE_ENLARGED_TYPE]: CurvedTileEnlargedModel
};

/**
 * Creates the tile model with respect to the given type.
 * @param {TileSpecifications} specs - The specifications for the tiles.
 * @param {string} type - The type of tile.
 * @param {string} direction - The direction of the tile.
 * @param {number} ratio - The size ratio of the tile.
 * @returns {TileModel} - Returns a tile model of the expected type.
 * @throws {TypeError} - If the given specifications object is not valid.
 * @throws {TypeError} - If the given type is not valid.
 * @throws {TypeError} - If the given direction is not valid.
 * @private
 */
export default (specs, type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) => {
    validateType(type);
    return new modelsMap[type](specs, direction, ratio);
};
