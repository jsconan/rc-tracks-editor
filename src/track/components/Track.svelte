<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { Sketch } from '../elements';
    import { Tile, TileNavigator } from '../../tile/components';
    import { TrackModel } from '../models';
    import { getRect, uid } from '../../core/helpers';

    export let track;
    export let x = 0;
    export let y = 0;
    export let width = void 0;
    export let height = void 0;

    TrackModel.validateInstance(track);

    const tilesStore = track.tilesStore;
    const modelsStore = track.modelsStore;

    const trackId = uid();
    const getId = id => `${trackId}-${id}`;

    let hoveredIndex = -1;
    const enter = event => {
        hoveredIndex = track.getIndex(event.target.dataset.id);
    };
    const leave = () => {
        hoveredIndex = -1;
    };

    $: rect = getRect($tilesStore);
</script>

<Sketch
    {x}
    {y}
    {width}
    {height}
    viewX={$tilesStore.x}
    viewY={$tilesStore.y}
    viewWidth={$tilesStore.width}
    viewHeight={$tilesStore.height}
>
    <TileNavigator elements={$tilesStore.tiles} {hoveredIndex} {...rect} on:select>
        {#each $tilesStore.tiles as { id, x, y, angle, model } (id)}
            <use
                data-id={id}
                {x}
                {y}
                href="#{getId(model.modelId)}"
                transform={model.getRotateTransform(x, y, angle)}
                role="menuitem"
                tabindex="-1"
                on:mouseenter={enter}
                on:mouseleave={leave}
            />
        {/each}
    </TileNavigator>
    <svelte:fragment slot="defs">
        {#each $modelsStore as { id, type, ratio, modelId } (id)}
            <Tile id={getId(modelId)} {type} {ratio} />
        {/each}
        <slot name="defs" />
    </svelte:fragment>
    <slot />
</Sketch>
