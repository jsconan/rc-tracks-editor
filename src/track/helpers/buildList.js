/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateAPI } from '../../core/helpers';
import { Vector2D } from '../../core/models';
import { TILE_DIRECTION_LEFT } from '../../tile/helpers';

/**
 * Process a list of tiles for rendering, computing the coordinates of each tile.
 * @param {*} list - The list of tiles from which build the track. It must implement a map method.
 * @param {object} config - A set of config options.
 * @param {number} [config.startX] - The X-coordinate of the first tile.
 * @param {number} [config.startY] - The Y-coordinate of the first tile.
 * @param {number} [config.tileAngle] - The rotation angle of each tile.
 * @param {number} [config.tileWidth] - The reserved width for each tile.
 * @param {number} [config.tileHeight] - The reserved height for each tile.
 * @param {number} [config.hPadding] - An horizontal padding added around the tiles.
 * @param {number} [config.vPadding] - A vertical padding added around the tiles.
 * @param {boolean} [config.centered] - Whether or not center the tiles.
 * @param {boolean} [config.aligned] - Whether or not align the tiles.
 * @param {boolean} [config.vertical] - Position the tiles vertically.
 * @returns {trackCoord} - Returns the list of coordinates.
 * @throws {TypeError} - If the given list does not implement the map method.
 */
export default (
    list,
    {
        startX = 0,
        startY = 0,
        tileAngle = 0,
        tileWidth = 0,
        tileHeight = 0,
        hPadding = 0,
        vPadding = 0,
        centered = false,
        aligned = false,
        vertical = false
    } = {}
) => {
    validateAPI(list, ['map']);

    const topLeft = new Vector2D();
    const bottomRight = new Vector2D();
    let tileX = startX;
    let tileY = startY;

    const tiles = list.map(model => {
        const curveAngle = model.getCurveAngle();
        const { id, type, direction, ratio } = model;
        const alignedAngle = curveAngle / 2;
        const directionAngle =
            direction === TILE_DIRECTION_LEFT ? Vector2D.STRAIGHT_ANGLE * 2 - alignedAngle : alignedAngle;
        const angle = tileAngle - (aligned && curveAngle < Vector2D.RIGHT_ANGLE ? directionAngle : 0);
        const rect = model.getBoundingRect(0, 0, angle);
        const width = Math.max(tileWidth, rect.width) + hPadding * 2;
        const height = Math.max(tileHeight, rect.height) + vPadding * 2;
        const dx = centered ? Math.max(0, tileWidth - rect.width) / 2 : 0;
        const dy = centered ? Math.max(0, tileHeight - rect.height) / 2 : 0;
        const topX = tileX - (centered && vertical ? width / 2 : 0) - hPadding;
        const topY = tileY - (!centered || vertical ? 0 : height / 2) - vPadding;
        const x = topX - rect.x + hPadding + dx;
        const y = topY - rect.y + vPadding + dy;

        topLeft.x = Math.min(topLeft.x, topX);
        topLeft.y = Math.min(topLeft.y, topY);

        bottomRight.x = Math.max(bottomRight.x, topX + width);
        bottomRight.y = Math.max(bottomRight.y, topY + height);

        if (vertical) {
            tileY += height;
        } else {
            tileX += width;
        }

        return { id, type, direction, ratio, x, y, angle, rect, model };
    });

    const { x, y } = topLeft;
    const { x: width, y: height } = bottomRight.sub(topLeft);

    return { x, y, width, height, tiles };
};

/**
 * @typedef {import('../models/TrackBuilder.js').trackCoord} trackCoord
 */
