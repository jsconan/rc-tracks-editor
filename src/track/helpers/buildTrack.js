/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateAPI } from '../../core/helpers';
import { Vector2D } from '../../core/models';

/**
 * Process a list of tiles for rendering a track, computing the coordinates of each tile.
 * @param {*} list - The list of tiles from which build the track. It must implement a map method.
 * @param {object} config - A set of config options.
 * @param {number} [config.startX] - The X-coordinate of the first tile.
 * @param {number} [config.startY] - The Y-coordinate of the first tile.
 * @param {number} [config.startAngle] - The rotation angle of the first tile.
 * @param {number} [config.hPadding] - An horizontal padding added around the track.
 * @param {number} [config.vPadding] - A vertical padding added around the tracks.
 * @returns {trackCoord} - Returns the list of coordinates.
 * @throws {TypeError} - If the given list does not implement the map method.
 */
export default (list, { startX = 0, startY = 0, startAngle = 0, hPadding = 0, vPadding = 0 } = {}) => {
    validateAPI(list, ['map']);

    const topLeft = new Vector2D();
    const bottomRight = new Vector2D();
    let inputX = startX;
    let inputY = startY;
    let inputAngle = startAngle;

    const tiles = list.map(model => {
        const rect = model.getBoundingRect(inputX, inputY, inputAngle);
        const { x, y, angle } = rect.input;
        const { id, type, direction, ratio } = model;

        inputX = rect.output.x;
        inputY = rect.output.y;
        inputAngle = rect.output.angle;

        topLeft.x = Math.min(topLeft.x, rect.x);
        topLeft.y = Math.min(topLeft.y, rect.y);

        bottomRight.x = Math.max(bottomRight.x, rect.x + rect.width);
        bottomRight.y = Math.max(bottomRight.y, rect.y + rect.height);

        return { id, type, direction, ratio, x, y, angle, rect, model };
    });

    const { x, y } = topLeft.subCoord(hPadding, vPadding);
    const { x: width, y: height } = bottomRight.sub(topLeft).addCoord(hPadding * 2, vPadding * 2);

    return { x, y, width, height, tiles };
};

/**
 * @typedef {object} trackCoord - Represents a track ready to be rendered.
 * @property {number} x - The left coordinate of the track.
 * @property {number} y - The top coordinate of the track.
 * @property {number} width - The width of the track.
 * @property {number} height - The height of the track.
 * @property {tileCoord[]} tiles - The list of tiles.
 */

/**
 * @typedef {object} tileCoord - Represents a positioned tile.
 * @property {string} id - The unique identifier of the tile.
 * @property {string} type - The type of tile.
 * @property {string} direction - The direction of the tile.
 * @property {number} ratio - The size ratio of the tile.
 * @property {number} x - The left coordinate of the tile.
 * @property {number} y - The top coordinate of the tile.
 * @property {number} angle - The rotation angle of the tile.
 * @property {TileModel} model - A reference to the tile model.
 */

/**
 * @typedef {import('../../tile/models').TileModel} TileModel
 */
