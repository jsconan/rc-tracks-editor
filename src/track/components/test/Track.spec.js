/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from '@testing-library/svelte';
import Track from '../Track.svelte';
import { TrackModel, TileSpecifications } from '../../models';
import { wait } from '../../../core/helpers';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;
const specs = new TileSpecifications(laneWidth, barrierWidth, barrierChunks);

describe('Track', () => {
    it('renders with default values', () => {
        const model = new TrackModel(specs);
        model.appendTile();
        const props = { model };
        const { container } = render(Track, { props });

        expect(container).toMatchSnapshot();
    });

    it('renders with the given parameters', () => {
        const model = new TrackModel(specs);
        model.appendTile();
        const props = {
            model,
            x: 100,
            y: 200,
            angle: 90
        };
        const { container } = render(Track, { props });

        expect(container).toMatchSnapshot();
    });

    it('update when the model is modified', async () => {
        const model = new TrackModel(specs);
        model.appendTile();
        const props = {
            model,
            x: 100,
            y: 200,
            angle: 90
        };
        const rendered = render(Track, { props });

        return wait(10)
            .then(() => model.appendTile())
            .then(() => wait(10))
            .then(() => expect(rendered.container).toMatchSnapshot());
    });

    it('needs a valid model', () => {
        expect(() =>
            render(Track, {
                props: {
                    model: {}
                }
            })
        ).toThrow("'model' is not a store with a 'subscribe' method");

        expect(() =>
            render(Track, {
                props: {
                    model: {
                        subscribe() {
                            return () => {};
                        }
                    }
                }
            })
        ).toThrow('The model must be an instance of TrackModel!');
    });
});
