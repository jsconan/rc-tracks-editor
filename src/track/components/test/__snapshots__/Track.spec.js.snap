// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track renders with default values 1`] = `
<body>
  <div>
    <svg
      viewBox="-155 0 200 210"
      x="0"
      y="0"
    >
      <defs />
      
      <g
        class="tile straight-tile"
        id="id-0"
        transform=""
      >
        <rect
          class="ground"
          height="110"
          width="90"
          x="-45"
          y="0"
        />
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-45"
            y="0"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-45"
            y="27.5"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-45"
            y="55"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-45"
            y="82.5"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="40"
            y="0"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="40"
            y="27.5"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="40"
            y="55"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="40"
            y="82.5"
          />
        </g>
      </g>
      
      
      <g
        class="tile curved-tile"
        id="id-1"
        transform=""
      >
        <path
          class="ground"
          d="M -45 110 A 10 10 0 0 1 -55 120 L -54.99999999999999 210 A 100 100 0 0 0 45 110 Z"
        />
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk even"
            d="M -45 110 A 10 10 0 0 1 -47.928932188134524 117.07106781186548 L -44.39339828220179 120.60660171779821 A 15 15 0 0 0 -40 110 Z"
          />
          
          <path
            class="barrier-chunk odd"
            d="M -47.928932188134524 117.07106781186548 A 10 10 0 0 1 -55 120 L -55 125 A 15 15 0 0 0 -44.39339828220179 120.60660171779821 Z"
          />
        </g>
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk odd"
            d="M 40 110 A 95 95 0 0 1 32.76855558857224 146.35492607468353 L 37.38795325112868 148.26834323650897 A 100 100 0 0 0 45 110 Z"
          />
          
          <path
            class="barrier-chunk even"
            d="M 32.76855558857224 146.35492607468353 A 95 95 0 0 1 12.175144212722017 177.175144212722 L 15.710678118654755 180.71067811865476 A 100 100 0 0 0 37.38795325112868 148.26834323650897 Z"
          />
          
          <path
            class="barrier-chunk odd"
            d="M 12.175144212722017 177.175144212722 A 95 95 0 0 1 -18.645073925316467 197.76855558857224 L -16.731656763491017 202.3879532511287 A 100 100 0 0 0 15.710678118654755 180.71067811865476 Z"
          />
          
          <path
            class="barrier-chunk even"
            d="M -18.645073925316467 197.76855558857224 A 95 95 0 0 1 -54.99999999999999 205 L -54.99999999999999 210 A 100 100 0 0 0 -16.731656763491017 202.3879532511287 Z"
          />
        </g>
      </g>
      
      
      <g
        class="tile curved-tile-enlarged"
        id="id-2"
        transform="rotate(90 -55 165)"
      >
        <path
          class="ground"
          d="M -100 165 A 10 10 0 0 1 -110 175 L -110 265 L -55 265 A 45 45 0 0 0 -10 220 L -10 165 Z"
        />
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk even"
            d="M -100 165 A 10 10 0 0 1 -102.92893218813452 172.07106781186548 L -99.39339828220179 175.6066017177982 A 15 15 0 0 0 -95 165 Z"
          />
          
          <path
            class="barrier-chunk odd"
            d="M -102.92893218813452 172.07106781186548 A 10 10 0 0 1 -110 175 L -110 180 A 15 15 0 0 0 -99.39339828220179 175.6066017177982 Z"
          />
        </g>
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk odd"
            d="M -15 220 A 40 40 0 0 1 -26.715728752538098 248.2842712474619 L -23.18019484660536 251.81980515339464 A 45 45 0 0 0 -10 220 Z"
          />
          
          <path
            class="barrier-chunk even"
            d="M -26.715728752538098 248.2842712474619 A 40 40 0 0 1 -55 260 L -55 265 A 45 45 0 0 0 -23.18019484660536 251.81980515339464 Z"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk even"
            height="5"
            width="27.5"
            x="-110"
            y="260"
          />
          
          <rect
            class="barrier-chunk odd"
            height="5"
            width="27.5"
            x="-82.5"
            y="260"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-15"
            y="165"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-15"
            y="192.5"
          />
        </g>
      </g>
      
      
    </svg>
    
  </div>
</body>
`;

exports[`Track renders with the given parameters 1`] = `
<body>
  <div>
    <svg
      height="400"
      viewBox="-220 -175 230 240"
      width="400"
      x="100"
      y="200"
    >
      <defs />
      
      <g
        class="tile straight-tile"
        id="id-0"
        transform="rotate(90 0 0)"
      >
        <rect
          class="ground"
          height="110"
          width="90"
          x="-45"
          y="0"
        />
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-45"
            y="0"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-45"
            y="27.5"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-45"
            y="55"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-45"
            y="82.5"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="40"
            y="0"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="40"
            y="27.5"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="40"
            y="55"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="40"
            y="82.5"
          />
        </g>
      </g>
      
      
      <g
        class="tile curved-tile"
        id="id-1"
        transform="rotate(90 -110 6.735557395310443e-15)"
      >
        <path
          class="ground"
          d="M -155 6.735557395310443e-15 A 10 10 0 0 1 -165 10.000000000000007 L -165 100 A 100 100 0 0 0 -65 6.735557395310443e-15 Z"
        />
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk even"
            d="M -155 6.735557395310443e-15 A 10 10 0 0 1 -157.92893218813452 7.071067811865482 L -154.3933982822018 10.606601717798219 A 15 15 0 0 0 -150 6.735557395310443e-15 Z"
          />
          
          <path
            class="barrier-chunk odd"
            d="M -157.92893218813452 7.071067811865482 A 10 10 0 0 1 -165 10.000000000000007 L -165 15.000000000000007 A 15 15 0 0 0 -154.3933982822018 10.606601717798219 Z"
          />
        </g>
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk odd"
            d="M -70 6.735557395310443e-15 A 95 95 0 0 1 -77.23144441142776 36.35492607468353 L -72.61204674887132 38.26834323650898 A 100 100 0 0 0 -65 6.735557395310443e-15 Z"
          />
          
          <path
            class="barrier-chunk even"
            d="M -77.23144441142776 36.35492607468353 A 95 95 0 0 1 -97.82485578727798 67.175144212722 L -94.28932188134524 70.71067811865474 A 100 100 0 0 0 -72.61204674887132 38.26834323650898 Z"
          />
          
          <path
            class="barrier-chunk odd"
            d="M -97.82485578727798 67.175144212722 A 95 95 0 0 1 -128.64507392531647 87.76855558857224 L -126.73165676349102 92.38795325112868 A 100 100 0 0 0 -94.28932188134524 70.71067811865474 Z"
          />
          
          <path
            class="barrier-chunk even"
            d="M -128.64507392531647 87.76855558857224 A 95 95 0 0 1 -165 95 L -165 100 A 100 100 0 0 0 -126.73165676349102 92.38795325112868 Z"
          />
        </g>
      </g>
      
      
      <g
        class="tile curved-tile-enlarged"
        id="id-2"
        transform="rotate(180 -165 -54.99999999999999)"
      >
        <path
          class="ground"
          d="M -210 -54.99999999999999 A 10 10 0 0 1 -220 -44.99999999999999 L -220 45.00000000000001 L -165 45.00000000000001 A 45 45 0 0 0 -120 7.105427357601002e-15 L -120 -54.99999999999999 Z"
        />
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk even"
            d="M -210 -54.99999999999999 A 10 10 0 0 1 -212.92893218813452 -47.92893218813452 L -209.3933982822018 -44.39339828220178 A 15 15 0 0 0 -205 -54.99999999999999 Z"
          />
          
          <path
            class="barrier-chunk odd"
            d="M -212.92893218813452 -47.92893218813452 A 10 10 0 0 1 -220 -44.99999999999999 L -220 -39.99999999999999 A 15 15 0 0 0 -209.3933982822018 -44.39339828220178 Z"
          />
        </g>
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk odd"
            d="M -125 7.105427357601002e-15 A 40 40 0 0 1 -136.7157287525381 28.284271247461906 L -133.18019484660536 31.819805153394643 A 45 45 0 0 0 -120 7.105427357601002e-15 Z"
          />
          
          <path
            class="barrier-chunk even"
            d="M -136.7157287525381 28.284271247461906 A 40 40 0 0 1 -165 40.00000000000001 L -165 45.00000000000001 A 45 45 0 0 0 -133.18019484660536 31.819805153394643 Z"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk even"
            height="5"
            width="27.5"
            x="-220"
            y="40.00000000000001"
          />
          
          <rect
            class="barrier-chunk odd"
            height="5"
            width="27.5"
            x="-192.5"
            y="40.00000000000001"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-125"
            y="-54.99999999999999"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-125"
            y="-27.499999999999993"
          />
        </g>
      </g>
      
      
    </svg>
    
  </div>
</body>
`;

exports[`Track update when the model is modified 1`] = `
<body>
  <div>
    <svg
      height="400"
      viewBox="-220 -175.00000000000003 230 240.00000000000003"
      width="400"
      x="100"
      y="200"
    >
      <defs />
      
      <g
        class="tile straight-tile"
        id="id-0"
        transform="rotate(90 0 0)"
      >
        <rect
          class="ground"
          height="110"
          width="90"
          x="-45"
          y="0"
        />
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-45"
            y="0"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-45"
            y="27.5"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-45"
            y="55"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-45"
            y="82.5"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="40"
            y="0"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="40"
            y="27.5"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="40"
            y="55"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="40"
            y="82.5"
          />
        </g>
      </g>
      
      
      <g
        class="tile curved-tile"
        id="id-1"
        transform="rotate(90 -110 6.735557395310443e-15)"
      >
        <path
          class="ground"
          d="M -155 6.735557395310443e-15 A 10 10 0 0 1 -165 10.000000000000007 L -165 100 A 100 100 0 0 0 -65 6.735557395310443e-15 Z"
        />
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk even"
            d="M -155 6.735557395310443e-15 A 10 10 0 0 1 -157.92893218813452 7.071067811865482 L -154.3933982822018 10.606601717798219 A 15 15 0 0 0 -150 6.735557395310443e-15 Z"
          />
          
          <path
            class="barrier-chunk odd"
            d="M -157.92893218813452 7.071067811865482 A 10 10 0 0 1 -165 10.000000000000007 L -165 15.000000000000007 A 15 15 0 0 0 -154.3933982822018 10.606601717798219 Z"
          />
        </g>
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk odd"
            d="M -70 6.735557395310443e-15 A 95 95 0 0 1 -77.23144441142776 36.35492607468353 L -72.61204674887132 38.26834323650898 A 100 100 0 0 0 -65 6.735557395310443e-15 Z"
          />
          
          <path
            class="barrier-chunk even"
            d="M -77.23144441142776 36.35492607468353 A 95 95 0 0 1 -97.82485578727798 67.175144212722 L -94.28932188134524 70.71067811865474 A 100 100 0 0 0 -72.61204674887132 38.26834323650898 Z"
          />
          
          <path
            class="barrier-chunk odd"
            d="M -97.82485578727798 67.175144212722 A 95 95 0 0 1 -128.64507392531647 87.76855558857224 L -126.73165676349102 92.38795325112868 A 100 100 0 0 0 -94.28932188134524 70.71067811865474 Z"
          />
          
          <path
            class="barrier-chunk even"
            d="M -128.64507392531647 87.76855558857224 A 95 95 0 0 1 -165 95 L -165 100 A 100 100 0 0 0 -126.73165676349102 92.38795325112868 Z"
          />
        </g>
      </g>
      
      
      <g
        class="tile curved-tile-enlarged"
        id="id-2"
        transform="rotate(180 -165 -54.99999999999999)"
      >
        <path
          class="ground"
          d="M -210 -54.99999999999999 A 10 10 0 0 1 -220 -44.99999999999999 L -220 45.00000000000001 L -165 45.00000000000001 A 45 45 0 0 0 -120 7.105427357601002e-15 L -120 -54.99999999999999 Z"
        />
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk even"
            d="M -210 -54.99999999999999 A 10 10 0 0 1 -212.92893218813452 -47.92893218813452 L -209.3933982822018 -44.39339828220178 A 15 15 0 0 0 -205 -54.99999999999999 Z"
          />
          
          <path
            class="barrier-chunk odd"
            d="M -212.92893218813452 -47.92893218813452 A 10 10 0 0 1 -220 -44.99999999999999 L -220 -39.99999999999999 A 15 15 0 0 0 -209.3933982822018 -44.39339828220178 Z"
          />
        </g>
        <g
          class="barrier curved-barrier"
        >
          
          <path
            class="barrier-chunk odd"
            d="M -125 7.105427357601002e-15 A 40 40 0 0 1 -136.7157287525381 28.284271247461906 L -133.18019484660536 31.819805153394643 A 45 45 0 0 0 -120 7.105427357601002e-15 Z"
          />
          
          <path
            class="barrier-chunk even"
            d="M -136.7157287525381 28.284271247461906 A 40 40 0 0 1 -165 40.00000000000001 L -165 45.00000000000001 A 45 45 0 0 0 -133.18019484660536 31.819805153394643 Z"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk even"
            height="5"
            width="27.5"
            x="-220"
            y="40.00000000000001"
          />
          
          <rect
            class="barrier-chunk odd"
            height="5"
            width="27.5"
            x="-192.5"
            y="40.00000000000001"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-125"
            y="-54.99999999999999"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-125"
            y="-27.499999999999993"
          />
        </g>
      </g>
      
      
      <g
        class="tile straight-tile"
        id="id-3"
        transform="rotate(270 -110 -110)"
      >
        <rect
          class="ground"
          height="110"
          width="90"
          x="-155"
          y="-110"
        />
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-155"
            y="-110"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-155"
            y="-82.5"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-155"
            y="-55"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-155"
            y="-27.5"
          />
        </g>
        <g
          class="barrier straight-barrier"
        >
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-70"
            y="-110"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-70"
            y="-82.5"
          />
          
          <rect
            class="barrier-chunk odd"
            height="27.5"
            width="5"
            x="-70"
            y="-55"
          />
          
          <rect
            class="barrier-chunk even"
            height="27.5"
            width="5"
            x="-70"
            y="-27.5"
          />
        </g>
      </g>
      
      
    </svg>
    
  </div>
</body>
`;
